main/ble_hidd_demo_main.c

```diff
@@ -496,15 +522,90 @@ void uart_parse_command (uint8_t character, struct cmdBuf * cmdBuffer)
             if(sec_conn == false) {
                 ESP_LOGI(EXT_UART_TAG,"not connected, cannot send report");
             } else {
+				/////// Keyboard
                 if (cmdBuffer->buf[1] == 0x00) {   // keyboard report
                     esp_hidd_send_keyboard_value(hid_conn_id,cmdBuffer->buf[0],&cmdBuffer->buf[2],6);
+                    
+                /////// Joystick
                 } else if (cmdBuffer->buf[1] == 0x01) {  // joystick report
-                    ESP_LOGI(EXT_UART_TAG,"joystick: buttons: 0x%X:0x%X:0x%X:0x%X",cmdBuffer->buf[2],cmdBuffer->buf[3],cmdBuffer->buf[4],cmdBuffer->buf[5]);
-                    //uint8_t joy[HID_JOYSTICK_IN_RPT_LEN];
-                    //memcpy(joy,&cmdBuffer->buf[2],HID_JOYSTICK_IN_RPT_LEN);
-                    ///@todo esp_hidd_send_joystick_value...
+					#if CONFIG_MODULE_USEJOYSTICK
+						///If we are using the rate limiter, check for the last transmission of a HID command.
+						///Note: as we sent absolute data for joystick axis, I think it is no problem to
+						/// discard any packets. If we get them too fast, many of them are still sent with the current values.
+						#if CONFIG_MODULE_USERATELIMITER
+							//check interval, jst to be sure use the absolute value of the subtraction.
+							if(abs(esp_timer_get_time() - cmdBuffer->lastTimeHIDCommandSent) > timeIntervalHIDCommand)
```

Comment:
is it guaranteed that the last (accumulated) data or button actions get sent if the rate limiter is active?
or can it happen that the last accumulated values are only sent when the next burst of mouse commands come in ?

Reply:
That's a good question. My intention was, that even if the rate limiter is active, most (but not all) commands will be sent.
I guess it is very unlikely, that e.g., the button release report is sent as last one of too much (only in this case it would trigger the rate limiter).

If we would implement any async mechanism for sending the last movement values and mouse buttons, it would increase the complexity quite much:
* New timer (I suggest a FreeRTOS timer)
* Callback structure
* Detecting if the timer should be started (how could we know in this command processor that this would be the last command?)

If we really need the proper transmission of every last movement value, I could implement it.
Nevertheless, I think if some external device hits the rate limiter, the chance of dropped values is quite high.

An easier approach:
If the rate limiter hits, just check if the button values have changed?

Reply:
I've added a button state tracker, which sends the button values if they change, even if the report is received out of the interval.

Reply:
i think that the rate limiter will always cause an unwanted behaviour if we do not guarantee that the most recent accumulated data is sent as soon as possible. for example: if movement data stays "in the pipeline" it will influence the next burst of mouse activities which come in, which might cause unintended deflection of the mouse cursor.
You could throw away acculumeted data if there was a pause, but IMO the only "clean" method to implement this is to trigger a timeout (a bit longer than the selected update rate) for a callback function where we check if accumulated coordinates are in the pipeline an send them if necessary.

Reply:
Good point, I will implement the accumulation in a different way:
* In `uart_parse_command` I will just accumulate the mouse values
* A timer (endless loop) will check the accumulated values on a regular timebase (-> use the rate limiter value) and
send any accumulated values.

do you think this is working :-)?

Reply:
sounds good :-)
in case the accumulator is empty, the report could be sent immediately in order to avoid unnecessary delay/jitter.